File: client/lib/serverUrl.js
// 서버 주소 정규화 + health 응답 판정 (phase 62 step0) — Electron·전역·파일시스템 비의존 순수 모듈.
// CRITICAL: 문자열 자르기·정규식으로 origin을 직접 조립하지 않는다 — 포트 생략·IPv6·유니코드 호스트에서
// 조용히 틀린 origin이 나오고, 그 값이 동일 출처 판정(= 새 창 허용 정책)의 입력이 된다. new URL()만 쓴다.
// 네트워크 호출은 이 모듈에 없다 — 프로브 실행은 step1 메인 프로세스 책임이고, 여기는 판정만 한다.

const SCHEME_RE = /^([a-zA-Z][a-zA-Z0-9+.-]*):/;
const ALLOWED_SCHEMES = new Set(['http', 'https']);

// 스킴 보정 — 운영자는 "192.168.0.10:3001"처럼 IP:포트만 친다.
// "localhost:3001"은 URL 문법상 스킴처럼 보이지만, 콜론 뒤가 포트 숫자(+경로/쿼리)면 host:port로 해석한다.
// 반환: 보정된 문자열 | null(허용 밖 스킴 — unsupported-scheme).
function ensureScheme(input) {
  const m = SCHEME_RE.exec(input);
  if (!m) return `http://${input}`;
  const scheme = m[1].toLowerCase();
  if (ALLOWED_SCHEMES.has(scheme)) return input;
  const rest = input.slice(m[0].length);
  if (/^\d+([/?#]|$)/.test(rest)) return `http://${input}`; // host:port 오인 케이스.
  return null;
}

export function normalizeServerUrl(input) {
  if (typeof input !== 'string') return { ok: false, reason: 'empty' };
  const trimmed = input.trim();
  if (!trimmed) return { ok: false, reason: 'empty' };
  const withScheme = ensureScheme(trimmed);
  if (withScheme === null) return { ok: false, reason: 'unsupported-scheme' };
  let url;
  try {
    url = new URL(withScheme);
  } catch {
    return { ok: false, reason: 'invalid' };
  }
  if (!ALLOWED_SCHEMES.has(url.protocol.replace(/:$/, ''))) return { ok: false, reason: 'unsupported-scheme' };
  // 자격증명이 설정 파일에 남는 표면을 만들지 않는다 — user@·user:pass@ 전부 거부.
  if (url.username || url.password) return { ok: false, reason: 'credentials' };
  if (!url.hostname) return { ok: false, reason: 'no-host' };
  // 경로·쿼리·해시는 버리고 origin만 — SPA 라우팅(/ → login.do)은 서버가 서빙하는 SPA 책임이다.
  return { ok: true, origin: url.origin };
}

export function healthUrl(origin) {
  return `${origin}/api/health`;
}

export function appUrl(origin) {
  return `${origin}/`;
}

export function isSameOrigin(url, origin) {
  try {
    return new URL(url).origin === new URL(origin).origin;
  } catch {
    return false; // 파싱 실패 = 동일 출처 아님(fail-closed).
  }
}

// 프로브 최종 origin 승격 판정 (phase 66 step4) — 리다이렉트를 따라간 응답의 최종 URL을 저장할 origin으로
// 승격할지 결정한다. fail-safe = 사용자가 친 주소 유지: 최종 URL이 없거나·문자열이 아니거나·파싱 불가·
// http/https 밖 스킴·자격증명 포함·https 입력의 http 하향이면 요청 origin 그대로다(자격증명 거부는 위
// 정규화 계약과 동형, 스킴 하향 거부는 사용자가 친 적 없는 평문 출처에 secure-origin 스위치가 걸리는
// 표면 차단 — 상향 http→https는 승격한다). 경로·쿼리·해시는 버린다 — 상단 CRITICAL대로 new URL만.
// 성공/실패 판정은 여기 없다 — 실패 시 승격 금지는 호출부(프로브)가 verdict로 가른다.
export function resolveFinalOrigin({ requestedOrigin, responseUrl } = {}) {
  const keep = { origin: requestedOrigin, changed: false };
  if (typeof responseUrl !== 'string') return keep;
  let url;
  try {
    url = new URL(responseUrl);
  } catch {
    return keep;
  }
  if (!ALLOWED_SCHEMES.has(url.protocol.replace(/:$/, ''))) return keep;
  if (url.username || url.password) return keep;
  if (url.protocol === 'http:' && isHttpsOrigin(requestedOrigin)) return keep; // 스킴 하향 미승격.
  return { origin: url.origin, changed: url.origin !== requestedOrigin };
}

function isHttpsOrigin(origin) {
  try {
    return new URL(origin).protocol === 'https:';
  } catch {
    return false; // 요청 origin이 파싱 불가면 하향 판정 불성립 — 기존 승격 규칙에 맡긴다.
  }
}

// 프로브 판정의 단일 출처(decisions (14)) — 200이라는 사실만으로는 부족하다.
// 임의의 웹 서버·사내 프록시·캡티브 포털도 200을 주므로, 본문이 { ok: true } JSON일 때만 인정한다.
// status가 null/undefined면 네트워크 자체가 실패한 경우다(호출부가 status를 못 얻음) → unreachable.
export function interpretHealthResponse({ status, body } = {}) {
  if (status === null || status === undefined) return { ok: false, reason: 'unreachable' };
  if (status !== 200) return { ok: false, reason: 'http-status' };
  const isPlainObject = typeof body === 'object' && body !== null && !Array.isArray(body);
  if (isPlainObject && body.ok === true) return { ok: true };
  return { ok: false, reason: 'not-article-server' };
}

File: client-qt/src/shell/serverurl.cpp
#include "shell/serverurl.h"

#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QLatin1Char>
#include <QLatin1String>
#include <QRegularExpression>
#include <QRegularExpressionMatch>
#include <QUrl>

namespace shell {
namespace {

// client/lib/serverUrl.js:6 - the scheme sniffer, character class for character class.
const QRegularExpression &schemeRe()
{
    static const QRegularExpression re(QStringLiteral("^([a-zA-Z][a-zA-Z0-9+.-]*):"));
    return re;
}

// client/lib/serverUrl.js:18 - the "host:port" heuristic (/^\d+([/?#]|$)/). [0-9] is spelled
// out on purpose: JS \d is ASCII-only, while PCRE \d would start matching other digit scripts
// the moment somebody adds UseUnicodePropertiesOption here.
const QRegularExpression &hostPortRestRe()
{
    static const QRegularExpression re(QStringLiteral("^[0-9]+([/?#]|$)"));
    return re;
}

bool isAllowedScheme(const QString &scheme)
{
    return scheme == QLatin1String("http") || scheme == QLatin1String("https");
}

int defaultPortFor(const QString &scheme)
{
    if (scheme == QLatin1String("http"))
        return 80;
    if (scheme == QLatin1String("https"))
        return 443;
    return -1;  // unknown scheme: no port is implied, so an explicit one is always spelled out
}

// Origin assembly, the half of parseUrlParts() this file was built around (X3). QUrl has no
// origin accessor, so the spelling is built by hand exactly once: every entry point here and
// in shell/diag.cpp goes through it. A second builder would make the same origin come out two
// ways and split the same-origin decision.
//
// Contract: scheme (lowercase) + "://" + host (IPv6 keeps its brackets) + ":" port, the port
// only when it differs from the scheme default. Returns false when the string is not a usable
// absolute URL WITH a host - that is this port's replacement for `new URL()` throwing (X2-a).
bool parsedOrigin(const QString &text, QString *origin, QString *scheme)
{
    const UrlParts parts = parseUrlParts(text);
    if (!parts.ok || !parts.hasHost)
        return false;
    if (origin)
        *origin = parts.origin;
    if (scheme)
        *scheme = parts.scheme;
    return true;
}

// client/lib/serverUrl.js:36 + :74 - one predicate for both credential checks, so the
// normalisation contract and the redirect contract can never drift apart.
bool hasCredentials(const QUrl &url)
{
    return !url.userName().isEmpty() || !url.password().isEmpty();
}

// client/lib/serverUrl.js:79-85 - fail-open-to-general-rule: if the requested origin does not
// parse, the "no downgrade" test simply does not apply (the general promotion rule wins).
// Do NOT turn this into fail-closed: it is the one helper in this file that leans open.
bool isHttpsOrigin(const QString &origin)
{
    QString scheme;
    if (!parsedOrigin(origin, nullptr, &scheme))
        return false;
    return scheme == QLatin1String("https");
}

NormalizedUrl rejected(const QString &reason)
{
    NormalizedUrl result;
    result.ok = false;
    result.reason = reason;
    return result;
}

// client/lib/serverUrl.js:12-20 - scheme repair. Operators type "192.168.0.10:3001", and
// "localhost:3001" looks like a scheme to every URL parser (QUrl reads it exactly the same
// way), so a numeric rest after the colon means host:port. Returns false for a scheme that is
// neither allowed nor a host:port mis-read.
bool ensureScheme(const QString &input, QString *out)
{
    const QRegularExpressionMatch match = schemeRe().match(input);
    if (!match.hasMatch()) {
        *out = QStringLiteral("http://") + input;
        return true;
    }
    if (isAllowedScheme(match.captured(1).toLower())) {
        *out = input;
        return true;
    }
    const QString rest = input.mid(match.capturedLength(0));
    if (hostPortRestRe().match(rest).hasMatch()) {
        *out = QStringLiteral("http://") + input;  // host:port mis-read, prefix the whole input
        return true;
    }
    return false;
}

} // namespace

UrlParts parseUrlParts(const QString &text)
{
    UrlParts parts;
    const QUrl url(text, QUrl::StrictMode);
    if (!url.isValid())
        return parts;

    parts.scheme = url.scheme().toLower();
    if (parts.scheme.isEmpty())
        return parts;  // relative reference ("garbage", "not a url", "", "/api/articles/5")
    parts.ok = true;
    parts.path = url.path(QUrl::FullyEncoded);

    // FullyEncoded keeps an internationalised host in its ACE/punycode form, which is what
    // url.origin gives in the canonical (measured: http://<hangul>.com -> xn--bj0bj06e.com).
    QString host = url.host(QUrl::FullyEncoded);
    if (host.isEmpty())
        return parts;  // "http://", "http://:3001", "mailto:user@h", "data:text/plain,x"
    parts.hasHost = true;
    if (host.contains(QLatin1Char(':')) && !host.startsWith(QLatin1Char('[')))
        host = QStringLiteral("[") + host + QStringLiteral("]");  // IPv6 literal

    parts.origin = parts.scheme + QStringLiteral("://") + host;
    const int port = url.port();  // -1 when absent; 0 is a real, explicitly written port
    if (port != -1 && port != defaultPortFor(parts.scheme)) {
        parts.origin += QLatin1Char(':');
        parts.origin += QString::number(port);
    }
    return parts;
}

NormalizedUrl normalizeServerUrl(const QString &input)
{
    const QString trimmed = input.trimmed();
    if (trimmed.isEmpty())
        return rejected(QStringLiteral("empty"));

    QString withScheme;
    if (!ensureScheme(trimmed, &withScheme))
        return rejected(QStringLiteral("unsupported-scheme"));

    const QUrl url(withScheme, QUrl::StrictMode);

    // Defence in depth (canonical R8): unreachable, because ensureScheme() either kept an
    // http/https input or prefixed one. Kept anyway - the predicate it shares with
    // resolveFinalOrigin is what the ftp/file/data/chrome-error rows exercise.
    if (url.isValid() && !isAllowedScheme(url.scheme().toLower()))
        return rejected(QStringLiteral("unsupported-scheme"));

    // Credentials are checked before the origin is built: "http://user@h:3001" parses fine,
    // and the point is to keep credentials out of the stored address entirely.
    if (url.isValid() && hasCredentials(url))
        return rejected(QStringLiteral("credentials"));

    QString origin;
    if (!parsedOrigin(withScheme, &origin, nullptr))
        return rejected(QStringLiteral("invalid"));

    NormalizedUrl result;
    result.ok = true;
    result.origin = origin;  // path, query and fragment are gone by construction
    return result;
}

QString healthUrl(const QString &origin)
{
    return origin + QStringLiteral("/api/health");
}

bool isSameOrigin(const QString &url, const QString &origin)
{
    QString left;
    QString right;
    if (!parsedOrigin(url, &left, nullptr))
        return false;  // fail-closed
    if (!parsedOrigin(origin, &right, nullptr))
        return false;
    return left == right;
}

FinalOrigin resolveFinalOrigin(const QString &requestedOrigin,
                               const std::optional<QString> &responseUrl)
{
    const FinalOrigin keep{ requestedOrigin, false };
    if (!responseUrl.has_value())
        return keep;

    QString origin;
    QString scheme;
    if (!parsedOrigin(*responseUrl, &origin, &scheme))
        return keep;
    if (!isAllowedScheme(scheme))
        return keep;
    if (hasCredentials(QUrl(*responseUrl, QUrl::StrictMode)))
        return keep;
    // Directional on purpose: http -> https is promoted two lines below, https -> http is not.
    // Making this symmetric re-opens the hole it closes (a captive portal or MITM proxy
    // flipping a stored https origin to plaintext).
    if (scheme == QLatin1String("http") && isHttpsOrigin(requestedOrigin))
        return keep;

    return FinalOrigin{ origin, origin != requestedOrigin };
}

HealthVerdict interpretHealthResponse(int status, const QByteArray &body)
{
    if (status < 0)
        return { false, QStringLiteral("unreachable") };
    if (status != 200)
        return { false, QStringLiteral("http-status") };

    // 200 alone proves nothing - any web server, corporate proxy or captive portal answers
    // 200. Only a JSON object whose "ok" is the boolean true is the article server.
    const QJsonDocument doc = QJsonDocument::fromJson(body);
    if (doc.isObject()) {
        const QJsonValue ok = doc.object().value(QStringLiteral("ok"));
        if (ok.isBool() && ok.toBool())
            return { true, QString() };
    }
    return { false, QStringLiteral("not-article-server") };
}

} // namespace shell
